File: packages/core/src/stats/signal.ts
/**
 * Signal processing — window functions, spectral estimation, smoothing and
 * correlation. Pure array→array, like the indicators: no plotting, no state.
 *
 * `fft` / `spectrogram` live in `stats/index.ts`; this module adds the pieces a
 * real analysis needs around them (a proper window, an averaged PSD, a filter).
 */
import { fft } from "./index.js";

/** Window (taper) applied to a frame before an FFT, to suppress spectral leakage. */
export type WindowName = "rectangular" | "hann" | "hamming" | "blackman" | "bartlett";

/**
 * Sample a window function over `n` points.
 *
 * `hann` is the sane default; `hamming` trades a touch of leakage for a lower
 * first sidelobe; `blackman` suppresses sidelobes hardest at the cost of
 * resolution; `bartlett` (triangular) is cheap; `rectangular` is no window at
 * all — use it only when the frame already contains whole periods.
 */
export function windowFunction(name: WindowName, n: number): Float64Array {
  const w = new Float64Array(n);
  if (n === 0) return w;
  if (n === 1) {
    w[0] = 1;
    return w;
  }
  const d = n - 1;
  for (let i = 0; i < n; i++) {
    const t = (2 * Math.PI * i) / d;
    switch (name) {
      case "rectangular": w[i] = 1; break;
      case "hamming": w[i] = 0.54 - 0.46 * Math.cos(t); break;
      case "blackman": w[i] = 0.42 - 0.5 * Math.cos(t) + 0.08 * Math.cos(2 * t); break;
      case "bartlett": w[i] = 1 - Math.abs((i - d / 2) / (d / 2)); break;
      default: w[i] = 0.5 - 0.5 * Math.cos(t); break; // hann
    }
  }
  return w;
}

/** A one-sided power spectral density estimate. */
export interface Psd {
  /** Frequencies in Hz (or cycles/sample when `sampleRate` is 1). */
  frequencies: Float64Array;
  /** Power per bin, in units²/Hz. */
  power: Float64Array;
}

export interface WelchOptions {
  /** Samples per segment (rounded down to a power of two). Default 256. */
  segment?: number;
  /** Overlap between segments, 0..1. Default 0.5. */
  overlap?: number;
  window?: WindowName;
  sampleRate?: number;
}

/**
 * Welch's method: average the periodograms of overlapping windowed segments.
 * Trades frequency resolution for a far lower-variance estimate than a single
 * FFT of the whole signal — the standard way to read a noisy spectrum.
 */
export function welch(signal: ArrayLike<number>, opts: WelchOptions = {}): Psd {
  const sr = opts.sampleRate ?? 1;
  const requested = opts.segment ?? 256;
  // The radix-2 FFT needs a power-of-two frame.
  let size = 1;
  while (size * 2 <= Math.min(requested, signal.length)) size *= 2;
  const bins = size >> 1;
  if (bins < 1) return { frequencies: new Float64Array(0), power: new Float64Array(0) };

  const overlap = Math.min(0.95, Math.max(0, opts.overlap ?? 0.5));
  const step = Math.max(1, Math.round(size * (1 - overlap)));
  const win = windowFunction(opts.window ?? "hann", size);
  // Normalization so the estimate is a density independent of the window shape.
  let winPower = 0;
  for (let i = 0; i < size; i++) winPower += win[i]! * win[i]!;
  const norm = 1 / (sr * winPower);

  const power = new Float64Array(bins);
  const re = new Float64Array(size);
  const im = new Float64Array(size);
  let segments = 0;
  for (let start = 0; start + size <= signal.length; start += step) {
    for (let i = 0; i < size; i++) {
      re[i] = signal[start + i]! * win[i]!;
      im[i] = 0;
    }
    fft(re, im);
    for (let b = 0; b < bins; b++) {
      // Double every bin except DC to fold the negative frequencies in.
      const scale = b === 0 ? 1 : 2;
      power[b]! += (re[b]! * re[b]! + im[b]! * im[b]!) * norm * scale;
    }
    segments++;
  }
  if (segments > 1) for (let b = 0; b < bins; b++) power[b]! /= segments;

  const frequencies = new Float64Array(bins);
  for (let b = 0; b < bins; b++) frequencies[b] = (b * sr) / size;
  return { frequencies, power };
}

/**
 * Savitzky–Golay smoothing: least-squares fit a polynomial over a sliding
 * window. Unlike a moving average it preserves peak height and width, which is
 * why spectroscopy and sensor pipelines reach for it.
 *
 * `window` must be odd (rounded up) and larger than `order`. Edges are handled
 * by clamping to the nearest sample.
 */
export function savitzkyGolay(values: ArrayLike<number>, window = 9, order = 2): Float64Array {
  const n = values.length;
  const out = new Float64Array(n);
  let m = Math.max(3, Math.floor(window));
  if (m % 2 === 0) m += 1;
  const half = (m - 1) >> 1;
  const deg = Math.max(0, Math.min(order, m - 1));
  if (n === 0) return out;
  if (n < m) {
    for (let i = 0; i < n; i++) out[i] = values[i]!;
    return out;
  }

  // Normal equations for the polynomial fit at the window centre (t = 0), which
  // is the only coefficient we need: out[i] = Σ c[k]·x[i+k].
  const size = deg + 1;
  const ata: number[][] = Array.from({ length: size }, () => new Array<number>(size).fill(0));
  for (let t = -half; t <= half; t++) {
    const powers = new Array<number>(size);
    powers[0] = 1;
    for (let p = 1; p < size; p++) powers[p] = powers[p - 1]! * t;
    for (let r = 0; r < size; r++) for (let c = 0; c < size; c++) ata[r]![c]! += powers[r]! * powers[c]!;
  }
  // Solve ATA·a = e0 by Gaussian elimination; the first row of ATA⁻¹ gives the weights.
  const aug: number[][] = ata.map((row, r) => [...row, r === 0 ? 1 : 0]);
  for (let c = 0; c < size; c++) {
    let pivot = c;
    for (let r = c + 1; r < size; r++) if (Math.abs(aug[r]![c]!) > Math.abs(aug[pivot]![c]!)) pivot = r;
    if (Math.abs(aug[pivot]![c]!) < 1e-12) continue;
    [aug[c], aug[pivot]] = [aug[pivot]!, aug[c]!];
    const p = aug[c]![c]!;
    for (let k = c; k <= size; k++) aug[c]![k]! /= p;
    for (let r = 0; r < size; r++) {
      if (r === c) continue;
      const f = aug[r]![c]!;
      if (f === 0) continue;
      for (let k = c; k <= size; k++) aug[r]![k]! -= f * aug[c]![k]!;
    }
  }
  const coef = aug.map((row) => row[size]!);

  const weights = new Float64Array(m);
  for (let t = -half, i = 0; t <= half; t++, i++) {
    let w = 0;
    let tp = 1;
    for (let p = 0; p < size; p++) {
      w += coef[p]! * tp;
      tp *= t;
    }
    weights[i] = w;
  }

  for (let i = 0; i < n; i++) {
    let acc = 0;
    for (let k = 0; k < m; k++) {
      const idx = Math.min(n - 1, Math.max(0, i + k - half));
      acc += weights[k]! * values[idx]!;
    }
    out[i] = acc;
  }
  return out;
}

/** A correlation sequence indexed by lag. */
export interface Correlation {
  /** Lags, from `-maxLag` to `+maxLag`. */
  lags: Int32Array;
  /** Correlation at each lag. */
  values: Float64Array;
}

/**
 * Cross-correlation of two series over ±`maxLag`. `normalize` (default true)
 * divides by the zero-lag autocorrelations, so the result is a correlation
 * coefficient in −1..1 and the peak lag reads directly as "b lags a by k".
 * Pass the same array twice for an autocorrelation.
 */
export function crossCorrelate(
  a: ArrayLike<number>,
  b: ArrayLike<number>,
  maxLag = Math.min(a.length, b.length) - 1,
  normalize = true,
): Correlation {
  const n = Math.min(a.length, b.length);
  const lag = Math.max(0, Math.min(Math.floor(maxLag), n - 1));
  let meanA = 0;
  let meanB = 0;
  for (let i = 0; i < n; i++) {
    meanA += a[i]!;
    meanB += b[i]!;
  }
  meanA /= n || 1;
  meanB /= n || 1;

  let varA = 0;
  let varB = 0;
  for (let i = 0; i < n; i++) {
    varA += (a[i]! - meanA) ** 2;
    varB += (b[i]! - meanB) ** 2;
  }
  const denom = normalize ? Math.sqrt(varA * varB) || 1 : 1;

  const count = 2 * lag + 1;
  const lags = new Int32Array(count);
  const values = new Float64Array(count);
  for (let k = -lag, j = 0; k <= lag; k++, j++) {
    let acc = 0;
    for (let i = 0; i < n; i++) {
      const bi = i + k;
      if (bi < 0 || bi >= n) continue;
      acc += (a[i]! - meanA) * (b[bi]! - meanB);
    }
    lags[j] = k;
    values[j] = acc / denom;
  }
  return { lags, values };
}

File: native/src/stats/signal.cpp
#include "stats/signal.hpp"

#include <algorithm>
#include <cmath>

#include "stats/stats.hpp"

namespace photon::stats {

namespace {
constexpr double kPi = 3.14159265358979323846;
}  // namespace

std::vector<double> window_function(Window name, size_t n) {
  std::vector<double> w(n, 0.0);
  if (n == 0) return w;
  if (n == 1) {
    w[0] = 1.0;
    return w;
  }
  const double d = static_cast<double>(n - 1);
  for (size_t i = 0; i < n; ++i) {
    const double t = 2.0 * kPi * static_cast<double>(i) / d;
    switch (name) {
      case Window::Rectangular:
        w[i] = 1.0;
        break;
      case Window::Hamming:
        w[i] = 0.54 - 0.46 * std::cos(t);
        break;
      case Window::Blackman:
        w[i] = 0.42 - 0.5 * std::cos(t) + 0.08 * std::cos(2.0 * t);
        break;
      case Window::Bartlett:
        w[i] = 1.0 - std::abs((static_cast<double>(i) - d / 2.0) / (d / 2.0));
        break;
      case Window::Hann:
      default:
        w[i] = 0.5 - 0.5 * std::cos(t);
        break;
    }
  }
  return w;
}

Psd welch(const double* signal, size_t count, int segment, double overlap, Window window,
          double sample_rate) {
  Psd out;
  if (!signal) count = 0;
  const size_t requested = segment > 0 ? static_cast<size_t>(segment) : 256;
  // The radix-2 FFT needs a power-of-two frame.
  size_t size = 1;
  while (size * 2 <= std::min(requested, count)) size *= 2;
  const size_t bins = size >> 1;
  if (bins < 1) return out;

  const double frac = std::min(0.95, std::max(0.0, overlap));
  const size_t step =
      std::max<size_t>(1, static_cast<size_t>(std::lround(static_cast<double>(size) * (1.0 - frac))));
  const std::vector<double> win = window_function(window, size);
  // Normalisation, so the estimate is a density independent of window shape.
  double win_power = 0.0;
  for (size_t i = 0; i < size; ++i) win_power += win[i] * win[i];
  const double norm = 1.0 / (sample_rate * win_power);

  out.power.assign(bins, 0.0);
  std::vector<double> re(size);
  std::vector<double> im(size);
  size_t segments = 0;
  for (size_t start = 0; start + size <= count; start += step) {
    for (size_t i = 0; i < size; ++i) {
      re[i] = signal[start + i] * win[i];
      im[i] = 0.0;
    }
    fft(re.data(), im.data(), size);
    for (size_t b = 0; b < bins; ++b) {
      // Every bin but DC doubles, to fold the negative frequencies in.
      const double scale = b == 0 ? 1.0 : 2.0;
      out.power[b] += (re[b] * re[b] + im[b] * im[b]) * norm * scale;
    }
    ++segments;
  }
  if (segments > 1) {
    for (size_t b = 0; b < bins; ++b) out.power[b] /= static_cast<double>(segments);
  }

  out.frequencies.resize(bins);
  for (size_t b = 0; b < bins; ++b) {
    out.frequencies[b] = static_cast<double>(b) * sample_rate / static_cast<double>(size);
  }
  return out;
}

std::vector<double> savitzky_golay(const double* values, size_t count, int window, int order) {
  std::vector<double> out(count, 0.0);
  if (!values || count == 0) return out;
  int m = std::max(3, window);
  if (m % 2 == 0) m += 1;
  const size_t width = static_cast<size_t>(m);
  const size_t half = (width - 1) / 2;
  const size_t size = static_cast<size_t>(std::max(0, std::min(order, m - 1))) + 1;
  if (count < width) {
    for (size_t i = 0; i < count; ++i) out[i] = values[i];
    return out;
  }

  // Normal equations for the polynomial fit at the window centre (t = 0). Only
  // the centre coefficient is needed, so the answer is one row of the inverse,
  // which is why the augmented column is e0 rather than a full identity.
  std::vector<double> aug(size * (size + 1), 0.0);
  std::vector<double> powers(size);
  for (int t = -static_cast<int>(half); t <= static_cast<int>(half); ++t) {
    powers[0] = 1.0;
    for (size_t p = 1; p < size; ++p) powers[p] = powers[p - 1] * static_cast<double>(t);
    for (size_t r = 0; r < size; ++r) {
      for (size_t c = 0; c < size; ++c) aug[r * (size + 1) + c] += powers[r] * powers[c];
    }
  }
  aug[0 * (size + 1) + size] = 1.0;

  for (size_t c = 0; c < size; ++c) {
    size_t pivot = c;
    for (size_t r = c + 1; r < size; ++r) {
      if (std::abs(aug[r * (size + 1) + c]) > std::abs(aug[pivot * (size + 1) + c])) pivot = r;
    }
    if (std::abs(aug[pivot * (size + 1) + c]) < 1e-12) continue;
    if (pivot != c) {
      for (size_t k = 0; k <= size; ++k) {
        std::swap(aug[c * (size + 1) + k], aug[pivot * (size + 1) + k]);
      }
    }
    const double p = aug[c * (size + 1) + c];
    for (size_t k = c; k <= size; ++k) aug[c * (size + 1) + k] /= p;
    for (size_t r = 0; r < size; ++r) {
      if (r == c) continue;
      const double f = aug[r * (size + 1) + c];
      if (f == 0.0) continue;
      for (size_t k = c; k <= size; ++k) {
        aug[r * (size + 1) + k] -= f * aug[c * (size + 1) + k];
      }
    }
  }

  std::vector<double> weights(width, 0.0);
  {
    size_t i = 0;
    for (int t = -static_cast<int>(half); t <= static_cast<int>(half); ++t, ++i) {
      double w = 0.0;
      double tp = 1.0;
      for (size_t p = 0; p < size; ++p) {
        w += aug[p * (size + 1) + size] * tp;
        tp *= static_cast<double>(t);
      }
      weights[i] = w;
    }
  }

  for (size_t i = 0; i < count; ++i) {
    double acc = 0.0;
    for (size_t k = 0; k < width; ++k) {
      const ptrdiff_t raw = static_cast<ptrdiff_t>(i + k) - static_cast<ptrdiff_t>(half);
      const size_t idx = static_cast<size_t>(
          std::min<ptrdiff_t>(static_cast<ptrdiff_t>(count) - 1, std::max<ptrdiff_t>(0, raw)));
      acc += weights[k] * values[idx];
    }
    out[i] = acc;
  }
  return out;
}

Correlation cross_correlate(const double* a, const double* b, size_t count, int max_lag,
                            bool normalize) {
  Correlation out;
  if (!a || !b || count == 0) return out;
  const int limit = static_cast<int>(count) - 1;
  const int lag = std::max(0, std::min(max_lag < 0 ? limit : max_lag, limit));

  double mean_a = 0.0;
  double mean_b = 0.0;
  for (size_t i = 0; i < count; ++i) {
    mean_a += a[i];
    mean_b += b[i];
  }
  mean_a /= static_cast<double>(count);
  mean_b /= static_cast<double>(count);

  double var_a = 0.0;
  double var_b = 0.0;
  for (size_t i = 0; i < count; ++i) {
    const double da = a[i] - mean_a;
    const double db = b[i] - mean_b;
    var_a += da * da;
    var_b += db * db;
  }
  double denom = 1.0;
  if (normalize) {
    denom = std::sqrt(var_a * var_b);
    if (denom == 0.0) denom = 1.0;
  }

  const size_t n = static_cast<size_t>(2 * lag + 1);
  out.lags.resize(n);
  out.values.resize(n);
  size_t j = 0;
  for (int k = -lag; k <= lag; ++k, ++j) {
    double acc = 0.0;
    for (size_t i = 0; i < count; ++i) {
      const ptrdiff_t bi = static_cast<ptrdiff_t>(i) + k;
      if (bi < 0 || bi >= static_cast<ptrdiff_t>(count)) continue;
      acc += (a[i] - mean_a) * (b[static_cast<size_t>(bi)] - mean_b);
    }
    out.lags[j] = k;
    out.values[j] = acc / denom;
  }
  return out;
}

}  // namespace photon::stats
